### DISCUSSION TITLE
快捷键扩展预设说明

### DISCUSSION THREAD

**Discussion by Geeyun-JY3** [the1812/Bilibili-Evolved#1646] [分享 / Share]
以下表格表示在本脚本中选择该预设后相应功能的快捷键。

| 播放器相关功能 | Default | YouTube  | [HTML5视频播放器增强脚本](https://github.com/xxxily/h5player) | PotPlayer |
| :-- | --- | --- | --- | --- |
| 播放／暂停视频 | <kbd>Space</kbd> | <kbd>k</kbd>/<kbd>Space</kbd> | <kbd>Space</kbd> | <kbd>Space</kbd> |
| 长快退<sup>1</sup> | <kbd>Shift</kbd>+<kbd>j</kbd> | <kbd>j</kbd> | <kbd>Ctrl</kbd>+<kbd>←</kbd> | <kbd>Ctrl</kbd>+<kbd>←</kbd> |
| 长快进<sup>1</sup> | <kbd>j</kbd> | <kbd>l</kbd> | <kbd>Ctrl</kbd>+<kbd>→</kbd> | <kbd>Ctrl</kbd>+<kbd>→</kbd> |
| 快进／快退 5 秒钟 | <kbd>←</kbd>/<kbd>→</kbd> | <kbd>←</kbd>/<kbd>→</kbd> | <kbd>←</kbd>/<kbd>→</kbd> | <kbd>←</kbd>/<kbd>→</kbd> |
| 上一帧 | <kbd>Shift</kbd>+<kbd>←</kbd> | <kbd>,</kbd> | <kbd>d</kbd> | <kbd>d</kbd> |
| 下一帧 | <kbd>Shift</kbd>+<kbd>→</kbd> | <kbd>.</kbd> | <kbd>f</kbd> | <kbd>f</kbd> |
| 减速播放 | <kbd>Shift</kbd>+<kbd>↓</kbd><br/>或<kbd>Shift</kbd>+<kbd>,</kbd> | <kbd>Shift</kbd>+<kbd>,</kbd> | <kbd>x</kbd> | <kbd>x</kbd> |
| 加速播放 | <kbd>Shift</kbd>+<kbd>↑</kbd><br/>或<kbd>Shift</kbd>+<kbd>.</kbd> | <kbd>Shift</kbd>+<kbd>.</kbd> | <kbd>c</kbd> | <kbd>c</kbd> |
| 正常／之前的倍速 | <kbd>Shift</kbd>+<kbd>/</kbd> | - | <kbd>z</kbd> | <kbd>z</kbd> |
| 将音量增大／减小 10% | <kbd>↑</kbd>/<kbd>↓</kbd> | <kbd>↑</kbd>/<kbd>↓</kbd> | <kbd>↑</kbd>/<kbd>↓</kbd> | <kbd>↑</kbd>/<kbd>↓</kbd> |
| 静音／取消静音 | <kbd>m</kbd> | <kbd>m</kbd> | ？（有相关 [issue](https://github.com/xxxily/h5player/issues/26) 但没试出来） | <kbd>m</kbd> |
| 跳至视频的开头 | <kbd>0</kbd> | <kbd>0</kbd>/<kbd>Home</kbd> | 同 Default | <kbd>Backspace</kbd> |
| 跳至视频的开头或结尾^ | #1613 | <kbd>Home</kbd>/<kbd>End</kbd> | - | - |
| 进入／退出全屏模式 | <kbd>f</kbd> | <kbd>f</kbd> | <kbd>Enter</kbd> | <kbd>Enter</kbd> |
| 进入／退出网页全屏 | <kbd>w</kbd> | 同 Default | <kbd>Shift</kbd>+<kbd>Enter</kbd> | <kbd>6</kbd> |
| 进入／退出影院模式（宽屏） | <kbd>t</kbd> | <kbd>t</kbd> | 同 Default | 同 Default |
| 画中画 | <kbd>p</kbd> | 同 Default | <kbd>Shift</kbd>+<kbd>p</kbd> | 同 Default |
| 上一个视频^ | <kbd>[</kbd> | <kbd>Shift</kbd>+<kbd>p</kbd> | - | <kbd>PgUp</kbd> |
| 下一个视频^ | <kbd>]</kbd> | <kbd>Shift</kbd>+<kbd>n</kbd> | <kbd>n</kbd> | <kbd>PgDn</kbd> |
| 未启用 `rememberVideoSpeed` 设置项时，显示当前倍速；<br/>启用 `rememberVideoSpeed` 设置项后，<br/>清除对某视频的倍数记忆，清除完成后视频速度将会回到后备值 | <kbd>Shift</kbd>+<kbd>;</kbd> | <kbd>Shift</kbd>+<kbd>;</kbd> | <kbd>Shift</kbd>+<kbd>;</kbd> | <kbd>Shift</kbd>+<kbd>;</kbd> |

| 特色功能 | Default | YouTube  | [HTML5视频播放器增强脚本](https://github.com/xxxily/h5player) | PotPlayer |
| --- | --- | --- | --- | --- |
| 搜索^ | #1197 | <kbd>/</kbd> | - | - |
| 截图 | <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>c</kbd>（不带弹幕）<br/><kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>Alt</kbd>+<kbd>c</kbd>（带弹幕） | 同 Default | <kbd>Shift</kbd>+<kbd>s</kbd> | 同 Default |
| 发弹幕 | <kbd>Enter</kbd> | 同 Default | 无 | 无 |
| 发表评论 | #1015 | <kbd>Ctrl</kbd>+<kbd>Enter</kbd> | 无 | 无 |
| 弹幕开关 | <kbd>d</kbd> | 同 Default | <kbd>Shift</kbd>+<kbd>d</kbd> | <kbd>Shift</kbd>+<kbd>d</kbd> |
| 打开播放器菜单 |<kbd>`</kbd> | 同 Default | 同 Default | 同 Default |
| 点赞（长按可一键三连，官方快捷键暂不支持） | <kbd>l</kbd>/<kbd>Shift</kbd>+<kbd>l</kbd> | <kbd>Shift</kbd>+<kbd>l</kbd> | 同 Default | 同 Default |
| 投币 | <kbd>c</kbd> | 同 Default | <kbd>Shift</kbd>+<kbd>c</kbd> | <kbd>Shift</kbd>+<kbd>c</kbd> |
| 收藏 | <kbd>s</kbd> | 同 Default | 同 Default | 同 Default |
| 稍后再看 | <kbd>Shift</kbd>+<kbd>w</kbd> | 同 Default | 同 Default | 同 Default |
| 快速收藏 | <kbd>Shift</kbd>+<kbd>s</kbd> | 同 Default | <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>s</kbd>              | 同 Default |

<sup>1</sup>: 长快进／快退时间可自定义

说明：
- 「^」表示该功能快捷键尚未实装
- 「同 Default」表示该预设对应网站／脚本／软件中无相对应功能，所以该预设快捷键同 Default 或 B 站原生快捷键。

<details>
<summary>已解决疑问</summary>

https://github.com/the1812/Bilibili-Evolved/issues/534#issuecomment-757092836 主要的作用是收集对比。现在实装几套预设后，再去看这个表格会有些迷惑的，而且我试了一下几套预设好像有些乱，感觉需要优化。感觉还是另开个 discussion 说明一下好一些，讨论完后我会编辑这个 discussion 把完整的表格放在这里（或者你们直接写进 doc 里？）。

有几个问题还是想问清楚。
1. 当 B 站（和本脚本）特色功能的快捷键（发弹幕、弹幕开关、点赞、投币、收藏、稍后再看、快速收藏）和启用预设的播放器相关功能快捷键重复时，是怎么处理的？我试了一下，如用 HTML5Player 和 PotPlayer 就没办法用 <kbd>c</kbd>、<kbd>d</kbd>、<kbd>Enter</kbd>  来投币、开关弹幕和发弹幕了，用 YouTube 就没办法用 <kbd>l</kbd> 来点赞了。但用 HTML5Player 时 <kbd>Shift</kbd>+<kbd>s</kbd> 又会同时触发快速收藏和截图。
2. B 站（和本脚本）原本有的功能，但当前启用预设对应网站／脚本／软件中无相对应功能（也就没有相应的快捷键），是否还有该快捷键呢？我的实测是如果本脚本 Default 预设的快捷键没和启用预设的快捷键重复，那就该功能的快捷键就会和本脚本 Default 预设的快捷键一致
3. @LonelySteve 能不能解释一下 Default 预设里 <kbd>Shift</kbd>+<kbd>;</kbd> 和 <kbd>Shift</kbd>+<kbd>/</kbd> 到底会触发什么？视频倍数相关功能没看懂🤦‍♂️

</details>

**Comment by Geeyun-JY3** [comment]
[功能文档](https://github.com/the1812/Bilibili-Evolved/blob/preview/doc/features.md#视频) 里快捷键扩展部分漏了个 Markdown 语法的括号
![image](https://user-images.githubusercontent.com/49396421/111058400-699b2400-84c9-11eb-99ee-831fc24a672b.png)

**Reply by the1812** [reply]
<img height="48" src="https://user-images.githubusercontent.com/26504152/111058685-fdb9bb00-84ca-11eb-808a-b98f4fb7e06f.png">

**Comment by the1812** [comment]
1. 预设会覆盖产生重复的快捷键, 同时触发快速收藏和截图是 bug, 应该只截图的.
2. 原有功能预设中没有的话就还是按 Default 里的

本质上预设也是一套自定义快捷键, 会覆盖到 Default 上面 (然后用户自定义快捷键会再来覆盖一次)
代码可以看这里: https://github.com/the1812/Bilibili-Evolved/blob/preview/src/video/keymap/key-binding-presets.ts

**Reply by Geeyun-JY3** [reply]
原来实现的方法这么简单粗暴……那就是说其实不会保证预设有 Default 的所有功能的快捷键

**Reply by the1812** [reply]
是的

**Comment by LonelySteve** [comment]
1. `Shift` + `/` 调用 VideoSpeedController 的 toggleVideoSpeed 方法，不传递参数，默认使用 `smart` 模式：

   https://github.com/the1812/Bilibili-Evolved/blob/4dc90004ffe49cf5b41727d1b5a56fd2edbc74f1/src/video/video-speed/video-speed-controller.ts#L259-L271

   首先，`classic` 模式是之前用的，很简单，就是切到上次**不同的**倍数，之所以强调**不同的**，是因为可能出现 1.0x, 1.25x, 1.25x, 1.5x, 1.5x 这样的倍数设置情况，你可以认为，中间重复的倍数没有意义。看成 1.0x, 1.25x, 1.5x 就行。当初这个模式使用的快捷键是独立的，可以在任意两个倍数之间来回切换。

   后来不是说要按一下 `Shift` + `/` 先重置倍数，再按一下恢复之前倍数嘛，这就是 `smart` 模式诞生的由来，说白了就是只在 1.0x 和其他倍数（也可能是 1.0x 自身）之前反复横跳，来回切换。

2. `Shift` + `;` 和记忆视频倍数功能有关，它存在的意义就是提供了一种快捷的手段去清除对当前视频的视频倍数记忆状态，因此，如果没有启用 `rememberVideoSpeed` 设置项（默认就是没有启用），这个快捷键就没有实际作用，只会和其他倍数相关的快捷键一样，触发一个当前倍数悬浮提示（这种情况下，你甚至可以认为它是一个倍数显示的快捷键，~~虽然它本意不是如此~~

   启用 `rememberVideoSpeed` 设置项，那么视频倍数记忆功能就可以细化到每个视频（通过 aid 区分），而不是仅仅靠用户选择的上次倍数。

   具体而言，可以参见以下代码：
  
   https://github.com/the1812/Bilibili-Evolved/blob/4dc90004ffe49cf5b41727d1b5a56fd2edbc74f1/src/video/video-speed/video-speed-controller.ts#L158

   1. 优先使用针对当前视频的特定记忆值
   2. 其次使用用户选择的上次倍数记忆值
   3. 最后使用分 P 共享倍数值（原生视频倍数记忆策略之一）

   只要用户开启了 `rememberVideoSpeed` 设置项，并且通过快捷键扩展或倍数菜单修改了当前视频的倍数，哪怕最后倍数修改回了最初值，仍然会新增一条对该视频的倍数记录，那么长久以往，就会越积越多，目前这个快捷键基本上算是应急设计，就算用户手滑改了视频的倍数，还能有个便捷的方式清掉记忆。

   显然，目前这个功能还不是很完善，因此默认是不启用的，并且没有在侧边栏中暴露出开关，不过我目前用着是没啥问题的。

**Reply by Geeyun-JY3** [reply]
简化一下就是：
<kbd>Shift</kbd>+<kbd>/</kbd> ：正常／之前的倍速

<kbd>Shift</kbd>+<kbd>;</kbd>：未启用 `rememberVideoSpeed` 设置项时，显示当前倍速；启用 `rememberVideoSpeed` 设置项后，清除对某视频的倍数记忆，清除完成后视频速度将会回到后备值

**Comment by Geeyun-JY3** [comment]
发现的 bug（[预览]v1.12.2 已修复）：三个预设的长快进长快退都搞反了（原网站/软件中的长快退按键在键盘的相对位置是左边，长快进是右边）

https://github.com/the1812/Bilibili-Evolved/issues/534#issuecomment-757092836 列出但没做进预设的快捷键：
1. （[预览]v1.12.2 已添加）YouTube 的 <kbd>k</kbd>：播放暂停
2. 上一个视频、下一个视频（三个预设都没有）

**Reply by the1812** [reply]
上一个和下一个还没做进快捷键扩展, Default 也没有的

**Reply by Geeyun-JY3** [reply]
哦也是，那是 B 站原生的

**Reply by Geeyun-JY3** [reply]
这下我懂为啥 https://github.com/the1812/Bilibili-Evolved/issues/1287#issuecomment-759972463 里说的三个都不会扩展到全网页范围了

**Reply by Geeyun-JY3** [reply]
就 TM 离谱，[YouTube 官方帮助页面](https://support.google.com/youtube/answer/7631406) 上/下一帧的快捷键居然写反了，搞得我之前 https://github.com/the1812/Bilibili-Evolved/issues/534#issuecomment-757092836 也跟着错了（脚本没问题，纯吐槽）

**Reply by Godwin9527** [reply]
> 上一个和下一个还没有做进键盘扩展，默认也没有的

有计划做进去吗？

**Reply by the1812** [reply]
> > 上一个和下一个还没有做进键盘扩展，默认也没有的
> 
> 有计划做进去吗？

没有, 这个操作没有对应的[播放器 API](https://github.com/the1812/Bilibili-Evolved/discussions/4341)

**Comment by Geeyun-JY3** [comment]
@the1812 刚刚编辑了一下，把完整的快捷键表格放在了这个 discussion 里。
看看有没啥问题，没问题的话建议从 release 引流到这里。

**Reply by Geeyun-JY3** [reply]
[v1.12.4](https://github.com/the1812/Bilibili-Evolved/releases/tag/v1.12.4) 的 release note 在之前更新日志的时候改回去了？目前又导回了 https://github.com/the1812/Bilibili-Evolved/issues/534#issuecomment-757092836

**Reply by the1812** [reply]
奇怪, 我都是直接复制的预览版, 应该不会改回去啊

**Comment by Geeyun-JY3** [comment]
官方加入了点赞的快捷键，是  <kbd>Shift</kbd>+<kbd>l</kbd>
![image](https://user-images.githubusercontent.com/49396421/119225233-1bdc1100-bb35-11eb-9fe6-ed50a207be01.png)

默认的预设的 <kbd>l</kbd> 是否要去掉？

**Reply by the1812** [reply]
那个不冲突倒没事, 预设里有些 <kbd>Shift + L</kbd> 的得去掉了

**Reply by Geeyun-JY3** [reply]
（好像也就一个吧
